### PR TITLE
[BEAM-3744] Expand Pubsub read API for Python.

### PR DESCRIPTION
**Pull request by udim**

### CODE REVIEW

**Comment by udim** [comment]
R: @chamikaramj

File: sdks/python/apache_beam/io/gcp/pubsub.py [hunk 1]
```diff
@@ -44,8 +68,8 @@
   pubsub_pb2 = None
 
 
-__all__ = ['PubsubMessage', 'ReadMessagesFromPubSub', 'ReadStringsFromPubSub',
-           'WriteStringsToPubSub']
+__all__ = ['PubsubMessage', 'ReadMessagesFromPubSub', 'ReadPayloadsFromPubSub',
```
Comment (aaltay) [inline sdks/python/apache_beam/io/gcp/pubsub.py]: Do we need 3 variations of ReadXFromPubSub transform? I would suggest considering making the underlying _ReadFromPubSub transform public, and letting users parametrize it to their needs.

File: sdks/python/apache_beam/io/gcp/pubsub.py [hunk 2]
```diff
@@ -118,97 +132,78 @@ class ReadMessagesFromPubSub(PTransform):
...
   def expand(self, pcoll):
     p = (pcoll.pipeline
          | _ReadFromPubSub(self.topic, self.subscription, self.id_label,
-                           with_attributes=False)
+                           with_attributes=False,
+                           timestamp_attribute=self.timestamp_attribute))
+    return p
+
+
+class ReadStringsFromPubSub(ReadPayloadsFromPubSub):
```
Comment (aaltay) [inline sdks/python/apache_beam/io/gcp/pubsub.py]: Do we actually need this transform? How is this handled in Java SDK?

File: sdks/python/apache_beam/io/gcp/pubsub.py (hunk 1 above)
Comment (udim) [reply sdks/python/apache_beam/io/gcp/pubsub.py]: There was a previous discussions about this here: https://github.com/apache/beam/pull/3223 .
Java has [PubsubIO.readStrings()](https://github.com/apache/beam/blob/29859eb54d05b96a9db477e7bb04537510273bd2/examples/java/src/main/java/org/apache/beam/examples/complete/game/StatefulTeamScore.java#L126), [readMessages()](https://github.com/apache/beam/blob/29859eb54d05b96a9db477e7bb04537510273bd2/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIO.java#L447), readMessagesWithAttributes(), etc.

File: sdks/python/apache_beam/io/gcp/pubsub.py (hunk 2 above)
Comment (udim) [reply sdks/python/apache_beam/io/gcp/pubsub.py]: Please see my comment above.

File: sdks/python/apache_beam/io/gcp/pubsub.py (hunk 1 above)
Comment (aaltay) [reply sdks/python/apache_beam/io/gcp/pubsub.py]: Thank you for the pointers.

I am trying to build a mental map of API available in both SDKs. You are proposing:

PTransforms:
- ReadMessagesFromPubSub: Outputs elements of type ``PubsubMessage``.
- ReadPayloadsFromPubSub: Outputs elements of type ``str``.
- ReadStringsFromPubSub: Outputs elements of type ``unicode``, decoded from UTF-8.

in Java there is
readMessages -> returns elements of type ``PubsubMessage.payload`
readMessagesWithAttributes -> returns both payload and attribute map
readStrings -> returns utf-8 decoded strings

There is also readProtos and readAvros.

Should we standardize on something like:
ReadStringsFromPubSub - strings
ReadMessagesFromPubSub - PubSubMessage (attributes being an argument to ReadMessagesFromPubSub, depending on that PubSubMessage will have the attributes or not.)

Thoughts?

File: sdks/python/apache_beam/io/gcp/pubsub.py [hunk 3]
```diff
@@ -280,25 +276,15 @@ def parse_subscription(full_subscription):
 class _PubSubSource(dataflow_io.NativeSource):
   """Source for the payload of a message as bytes from a Cloud Pub/Sub topic.
 
+  This ``PTransform`` is overridden by a native Pubsub implementation.
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/pubsub.py]: This is not a PTransform.

File: sdks/python/apache_beam/io/gcp/pubsub.py [hunk 4]
```diff
@@ -338,7 +325,7 @@ def display_data(self):
 
   def reader(self):
     raise NotImplementedError(
-        'PubSubPayloadSource is not supported in local execution.')
+        'PubSubSource is not supported in local execution.')
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/pubsub.py]: PubSubSource is an implementation detail. "Reading form Cloud Pub/Sub is not supported in local execution" ?

File: sdks/python/apache_beam/io/gcp/pubsub_test.py [hunk 5]
```diff
@@ -276,16 +283,27 @@ def topic(self, name):
     return FakePubsubTopic(name, self)
 
 
+def create_client_message(payload, message_id, attributes, publish_time):
+  """Returns a message as it would be returned from Pubsub client."""
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/pubsub_test.py]: Cloud Pub/Sub client

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py [hunk 6]
```diff
@@ -742,7 +742,7 @@ def run_Read(self, transform_node):
       standard_options = (
           transform_node.inputs[0].pipeline.options.view_as(StandardOptions))
       if not standard_options.streaming:
-        raise ValueError('PubSubPayloadSource is currently available for use '
+        raise ValueError('PubSubSource is currently available for use '
```
Comment (chamikaramj) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: Ditto.

File: sdks/python/apache_beam/runners/direct/transform_evaluator.py [hunk 7]
```diff
@@ -428,24 +428,41 @@ def _read_from_pubsub(self):
...
         max_messages=10) as results:
       def _get_element(message):
-        if self.source.with_attributes:
-          return PubsubMessage._from_message(message)
+        parsed_message = PubsubMessage._from_message(message)
+        if timestamp_attribute:
+          try:
+            rfc3339_or_milli = parsed_message.attributes[timestamp_attribute]
+          except KeyError:
+            raise KeyError('Timestamp attribute not found: %s' %
```
Comment (chamikaramj) [inline sdks/python/apache_beam/runners/direct/transform_evaluator.py]: Raise/preserve original exception ?

File: sdks/python/apache_beam/runners/direct/transform_evaluator.py [hunk 8]
```diff
@@ -428,24 +428,41 @@ def _read_from_pubsub(self):
...
+                           self.source.timestamp_attribute)
+          try:
+            timestamp = Timestamp.from_rfc3339(rfc3339_or_milli)
+          except ValueError:
+            try:
+              timestamp = Timestamp(micros=int(rfc3339_or_milli) * 1000)
+            except ValueError:
+              raise ValueError('Invalid timestamp value: %s', rfc3339_or_milli)
```
Comment (chamikaramj) [inline sdks/python/apache_beam/runners/direct/transform_evaluator.py]: Raise/preserve original exception ?

File: sdks/python/apache_beam/utils/timestamp.py [hunk 9]
```diff
@@ -38,6 +40,12 @@ class Timestamp(object):
   """
 
   def __init__(self, seconds=0, micros=0):
+    if not isinstance(seconds, (int, float, long)):
+      raise TypeError('Cannot interpret %s %s as seconds.' % (
+        seconds, type(seconds)))
```
Comment (chamikaramj) [inline sdks/python/apache_beam/utils/timestamp.py]: type is the first parameter ?

File: sdks/python/apache_beam/utils/timestamp.py [hunk 10]
```diff
@@ -38,6 +40,12 @@ class Timestamp(object):
...
+    if not isinstance(micros, (int, float, long)):
+      raise TypeError('Cannot interpret %s %s as micros.' % (
```
Comment (chamikaramj) [inline sdks/python/apache_beam/utils/timestamp.py]: Ditto.

File: sdks/python/apache_beam/utils/timestamp.py [hunk 11]
```diff
@@ -53,12 +61,52 @@ def of(seconds):
       Corresponding Timestamp object.
     """
 
-    if isinstance(seconds, Duration):
-      raise TypeError('Can\'t interpret %s as Timestamp.' % seconds)
+    if not isinstance(seconds, (int, float, Timestamp)):
+      raise TypeError('Cannot interpret %s %s as Timestamp.' % (
+        seconds, type(seconds)))
```
Comment (chamikaramj) [inline sdks/python/apache_beam/utils/timestamp.py]: Ditto.

File: sdks/python/apache_beam/io/gcp/pubsub.py (hunk 1 above)
Comment (chamikaramj) [reply sdks/python/apache_beam/io/gcp/pubsub.py]: @lukecwik @charlesccychen can you comment on whether we actually need 'ReadStringsFromPubSub' for some reason that we missed here ?

Comment (charlesccychen) [reply sdks/python/apache_beam/io/gcp/pubsub.py]: No, we don't need it. It was simply a tentative API choice at the time.

Comment (udim) [reply sdks/python/apache_beam/io/gcp/pubsub.py]: After talking to @aaltay this morning, I'm going to reduce the API to just 1 transform that does everything.

Comment (aaltay) [reply sdks/python/apache_beam/io/gcp/pubsub.py]: I maybe wrong. AFAIK, Dataflow implementation for reading only the payloads is faster that reading the whole message. I assume, we can still do this in DataflowRunner using transform overrides.

**Comment by chamikaramj** [comment]
Were the comments addressed ? Lemme know if I can continue the review.

**Comment by udim** [comment]
@chamikaramj Yes, please continue. I'll remove ReadStringsFromPubSub in a subsequent PR.

**Comment by udim** [comment]
@chamikaramj , I believe your comments are referring to a different PR?
retest this please

**Comment by chamikaramj** [comment]
Yeah, that was an invalid diff. Please ignore those comments. I'm re-reviewing.

**Comment by chamikaramj** [comment]
LGTM. Thanks.

**Comment by lukecwik** [comment]
@chamikaramj Were you planning to merge this?

**Comment by chamikaramj** [comment]
Yeah, merging now. Was waiting for tests to pass.

**Pull request merged**